### PR TITLE
Add support for broadcast `QubitUnitary` in `zyz_decomposition`

### LINKED ISSUE #2636: Implement decomposition of `QubitUnitary` with parameter broadcasting
### Feature details

When using a broadcasted unitary matrix in `QubitUnitary` (possible as of #2609), the `decomposition` is not defined yet.
This can be implemented but may not be entirely straightforward.

### Implementation

To implement this feature:
1. The broadcasted matrix of shape `(batch_size, dim, dim)` has to be split into `batch_size` matrices (trivial)
2. The correct decomposition method (`zyz_decomposition` for 1 qubit/`two_qubit_decomposition` for 2 qubits) has to be applied to each of the matrices from step 1. (for example list comprehension)
3. The data of the operations in the resulting decompositions has to be read out and unified (may be tricky depending on how homogeneous the decompositions obtained in step 2. are),
4. operations containing the broadcasted data from step 3 have to be collected into a single decomposition list (might also be a bit tricky).

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Shiro-Raven**
**Context:**
This is one of two PRs (the second being #3494) adding support for broadcast parameters to the unitary decomposition routine of the `QubitUnitary` operator.

**Description of the Change:**
- Users now can retrieve individual decompositions for each item in a broadcast single-qubit `QubitUnitary`. This used to raise an error, but now is allowed. An example usage is shown below:
```python
>>> mats = qml.RY.compute_matrix(np.array([1.2, 2.3, 2.7]))
>>> op = qml.QubitUnitary(mats, wires=0)
>>> decompositions = op.decomposition()
>>> pprint(decompositions)
[[Rot(tensor(-2.98266051e-128, requires_grad=True), tensor(1.2, requires_grad=True), tensor(5.59403882e-129, requires_grad=True), wires=[0])],
 [Rot(tensor(-3.5436292e-128, requires_grad=True), tensor(2.3, requires_grad=True), tensor(-1.35248216e-128, requires_grad=True), wires=[0])],
 [Rot(tensor(-5.59095128e-128, requires_grad=True), tensor(2.7, requires_grad=True), tensor(-3.54118996e-128, requires_grad=True), wires=[0])]]
```
- The approach relies on taking advantage of vectorized operations to maximize performance in comparison to a simple loop-and-gather approach.
- The already-existent behavior for non-batched input is still maintained 

**Benefits:**
- Performance gain for users who previously had to calculate decompositions individually and later on merge the results.

**Possible Drawbacks:**
The current implementation doesn't distinguish between the following two usages.
```python
# Usage 1
>>> mats = qml.RY.compute_matrix(np.array([1.2]))
>>> op = qml.QubitUnitary(mats, wires=0)
>>> decompositions = op.decomposition()
# Usage 2
>>> mats = qml.RY.compute_matrix(1.2)
>>> op = qml.QubitUnitary(mats, wires=0)
>>> decompositions = op.decomposition()
```
with the output in both bases having the form `[Rot(tensor(-2.98266051e-128, requires_grad=True), tensor(1.2, requires_grad=True), tensor(5.59403882e-129, requires_grad=True), wires=[0])]` (a single-item list). This can be a source of confusion to the users. A more consistent output may be a list of a list containing the decomposition for Usage 1, and a minor change of the code would be needed if others agree this is a better output. 

**Related GitHub Issues:**
Resolves #2636

### CODE REVIEW

**Comment by albi3ro** [comment]
We currently have a PR up adding more documentation about broadcasting: https://github.com/PennyLaneAI/pennylane/pull/3356/files

Sorry I didn't link this earlier. Hope it helps.

For example of what we are going for, consider:

```pycon
>>> qml.IsingXX(np.array([1.2, 2.3, 3.4]), wires=(0,1)).decomposition()
[CNOT(wires=[0, 1]),
 RX(tensor([1.2, 2.3, 3.4], requires_grad=True), wires=[0]),
 CNOT(wires=[0, 1])]
```

Since the `IsingXX` gate started with a broadcasted parameter, it's decomposition ends up with a `RX` gate with a broadcasted parameter.

Unfortunately, the stated interface for [`Operator.decomposition`](https://docs.pennylane.ai/en/stable/code/api/pennylane.operation.Operator.html#pennylane.operation.Operator.decomposition) is that it returns `list[Operator]`. Deviating from that would require a breaking change to an important method for a  core data structure.

For now, we can make the assumption in `zyz_decompsoition` that if the input matrix is broadcasted, we will always return a `Rot` gate. That way we don't have to worry about choosing between `Rot` and `RZ`.

**Comment by albi3ro** [comment]
The end goal is to support the decomposition of `QubitUnitary` when the matrix input has a batch dimension and (at least) acts on one wire.  This could be accomplished through purely changing `QubitUnitary.decomposition`, but it could also be accomplished by changing `zyz_decomposition`.

I shared the `IsingXX` example to show what a decomposition with broadcasted parameters should look like.

Instead of:
```
[[Rot(tensor(-2.98266051e-128, requires_grad=True), tensor(1.2, requires_grad=True), tensor(5.59403882e-129, requires_grad=True), wires=[0])],
 [Rot(tensor(-3.5436292e-128, requires_grad=True), tensor(2.3, requires_grad=True), tensor(-1.35248216e-128, requires_grad=True), wires=[0])],
 [Rot(tensor(-5.59095128e-128, requires_grad=True), tensor(2.7, requires_grad=True), tensor(-3.54118996e-128, requires_grad=True), wires=[0])]]
```

Something like:
```
>>> op.decomposition()
[Rot(tensor([0, 0, 0], requires_grad=True), tensor([1.2, 2.3, 2.7], requires_grad=True), tensor([0, 0, 0], requires_grad=True), wires=[0])]
```

File: pennylane/transforms/decompositions/single_qubit_unitary.py [hunk 1]
```diff
@@ -76,26 +76,60 @@ def zyz_decomposition(U, wire):
...
     # undefined case of 0/0. We add a smaller value to the imaginary part than
     # the real part because it is imag / real in the definition of atan2.
-    angle_U00 = math.arctan2(math.imag(U[0, 0]) + 1e-128, math.real(U[0, 0]) + 1e-64)
-    angle_U10 = math.arctan2(math.imag(U[1, 0]) + 1e-128, math.real(U[1, 0]) + 1e-64)
-
-    phi = -angle_U10 - angle_U00
-    omega = angle_U10 - angle_U00
-
-    return [qml.Rot(phi, theta, omega, wires=wire)]
+    epsilon_real, epsilon_imag = 1e-64, 1e-128
```
Comment (albi3ro) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:112]:
```suggestion
    epsilon = math.finfo(phi.dtype).eps
```
This was a problem with the old implementation, but it should have better value for machine epsilon and avoid the divide by zeros warnings.

Comment (Shiro-Raven) [reply pennylane/transforms/decompositions/single_qubit_unitary.py]: This change gave me issues when testing with the TensorFlow backend. But I found out that the imaginary epsilon was not necessary since we only need to ensure that the denominate (the real part), so I changed the code accordingly.

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -435,6 +435,30 @@
...
+  Note that no RZ decompositions are returned for broadcast operators:
+
+  ```python
+  >>> mats = qml.RZ.compute_matrix(np.array([1.2, 2.3, 2.7]))
+  >>> op = qml.QubitUnitary(mats, wires=0)
+  >>> decompositions = op.decomposition()
+  >>> print(decompositions)
+  [Rot(array([0.6 , 1.15, 1.35]), array([0., 0., 0.]), array([0.6 , 1.15, 1.35]), wires=[0])]
+  ```  
+
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:461]: For a change of this scale, we don't need an example.

File: pennylane/transforms/decompositions/single_qubit_unitary.py [hunk 3]
```diff
@@ -76,26 +76,39 @@ def zyz_decomposition(U, wire):
...
-    # Compute phi and omega from the angles of the top row; use atan2 to keep
-    # the angle within -np.pi and np.pi, and add very small values to avoid the
-    # undefined case of 0/0. We add a smaller value to the imaginary part than
-    # the real part because it is imag / real in the definition of atan2.
-    angle_U00 = math.arctan2(math.imag(U[0, 0]) + 1e-128, math.real(U[0, 0]) + 1e-64)
-    angle_U10 = math.arctan2(math.imag(U[1, 0]) + 1e-128, math.real(U[1, 0]) + 1e-64)
+    # If U is only one unitary and its value is not abstract, we can include a conditional
+    # statement that will check if the off-diagonal elements are 0; if so, just use one RZ
+    if len(U) == 1 and not math.is_abstract(U[0]):
+        if math.allclose(U[0, 0, 1], 0.0):
```
Comment (albi3ro) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:88]:
```suggestion
    if len(U) == 1 and not math.is_abstract(U) and math.allclose(U[0, 0, 1], 0.0):
```
An optional style change.  My local linter tends to recommend combining if statements when possible.

File: tests/ops/qubit/test_matrix_ops.py [hunk 4]
```diff
@@ -269,17 +269,6 @@ def test_qubit_unitary_decomposition(self, U, expected_gate, expected_params):
         assert isinstance(decomp2[0], expected_gate)
         assert np.allclose(decomp2[0].parameters, expected_params, atol=1e-7)
 
-    def test_error_qubit_unitary_decomposition_broadcasted(self):
-        """Tests that broadcasted QubitUnitary decompositions are not supported."""
-        U = np.array(
-            [[0.98877108 + 0.0j, 0.0 - 0.14943813j], [0.0 - 0.14943813j, 0.98877108 + 0.0j]]
-        )
-        U = np.tensordot([1j, -1.0, (1 + 1j) / np.sqrt(2)], U, axes=0)
```
Comment (albi3ro) [inline tests/ops/qubit/test_matrix_ops.py:277]: Since we're still raising this error for the two qubit case, we should just modify this test instead of delete it.  We need to make sure the error is still raised in the two-qubit case.

We can easily generate a two-qubit broadcasted unitary matrix doing:
```
U = qml.IsingYY.compute_matrix(np.array([1.2, 2.3, 3.4]))
```

File: tests/transforms/test_decompositions.py [hunk 5]
```diff
@@ -54,35 +55,57 @@
...
 
 class TestQubitUnitaryZYZDecomposition:
     """Test that the decompositions are correct."""
 
-    @pytest.mark.parametrize("U,expected_gate,expected_params", single_qubit_decomps)
-    def test_zyz_decomposition(self, U, expected_gate, expected_params):
-        """Test that a one-qubit matrix in isolation is correctly decomposed."""
-        obtained_gates = zyz_decomposition(U, Wires("a"))
-
+    def _run_assertions(self, U, expected_gate, expected_params, obtained_gates):
```
Comment (albi3ro) [inline tests/transforms/test_decompositions.py:80]: 👍 good abstraction of repeated code.

**Review by albi3ro — COMMENTED** [review]
Looking great! Thanks for the adjustment in the output form.

Remember to add your name to the contributors list at the bottom of the change log. 

I would also like to request a test checking the integration between `QubitUnitary.decomposition` and `zyz_decomposition`.  I see tests validating the behaviour for just `zyz_decomposition`, but we also need to check that `QubitUnitary.decomposition` calls it and returns correctly.

File: pennylane/transforms/decompositions/single_qubit_unitary.py (hunk 3 above)
Comment (Shiro-Raven) [reply pennylane/transforms/decompositions/single_qubit_unitary.py:88]: I can see how the brevity here is more guideline-compliant. But personally I believe this separation of the conditions for easier mental separation.

Comment (albi3ro) [reply pennylane/transforms/decompositions/single_qubit_unitary.py:88]: Sounds good.

File: tests/ops/qubit/test_matrix_ops.py [hunk 6]
```diff
@@ -269,16 +279,30 @@ def test_qubit_unitary_decomposition(self, U, expected_gate, expected_params):
...
+        ) as mocked_zyz_decomposition:
+            decomp = qml.QubitUnitary.compute_decomposition(H, wires=0)
+            np.testing.assert_equal(mocked_zyz_decomposition.call_args.args[0], H)
+            assert mocked_zyz_decomposition.call_args.args[1] == 0
+
+            decomp2 = qml.QubitUnitary(X, wires=1).decomposition()
+            np.testing.assert_equal(mocked_zyz_decomposition.call_args.args[0], X)
+            assert mocked_zyz_decomposition.call_args.args[1] == 1
+
+            assert mocked_zyz_decomposition.call_count == 2
```
Comment (albi3ro) [inline tests/ops/qubit/test_matrix_ops.py:296]: I don't think we need to test whether or not `zyz_decomposition` was actually called.

What we need to check is whether it gives the right answer, not necessarily "how" it got to that right answer.

While they can be helpful, I often find mocking and patching causes more trouble then they're worth outside of specific situations where we have no other options.

If we know `decomposition` gives the right answer, we are properly tested.

```suggestion
```

File: doc/releases/changelog-dev.md (hunk 2 above)
Comment (Shiro-Raven) [reply doc/releases/changelog-dev.md]: Removed :)

**Review by albi3ro — APPROVED** [review]
Excellent job @Shiro-Raven  🎉 I'm glad to be getting this improved broadcasting support.

Hope you enjoyed the process.
